src/decode.rs

```diff
@@ -3525,280 +3525,231 @@ unsafe fn decode_b(
+            let mut pal_idx;
             if t.frame_thread.pass != 0 {
-                let p_0 = t.frame_thread.pass & 1;
-                if (ts.frame_thread[p_0 as usize].pal_idx).is_null() {
-                    unreachable!();
-                }
-                pal_idx_0 = ts.frame_thread[p_0 as usize].pal_idx;
-                ts.frame_thread[p_0 as usize].pal_idx =
-                    (ts.frame_thread[p_0 as usize].pal_idx).offset((cbw4 * cbh4 * 16) as isize);
+                let p = t.frame_thread.pass & 1;
+                let frame_thread = &mut ts.frame_thread[p as usize];
+                assert!(!(frame_thread.pal_idx).is_null());
+                pal_idx = frame_thread.pal_idx;
+                frame_thread.pal_idx = frame_thread.pal_idx.offset((cbw4 * cbh4 * 16) as isize);
             } else {
-                pal_idx_0 = &mut *(t.scratch.c2rust_unnamed_0.pal_idx)
+                pal_idx = t
+                    .scratch
+                    .c2rust_unnamed_0
+                    .pal_idx
                     .as_mut_ptr()
-                    .offset((bw4 * bh4 * 16) as isize) as *mut uint8_t;
+                    .offset((bw4 * bh4 * 16) as isize);
             }
-            read_pal_indices(t, pal_idx_0, b, 1 as libc::c_int, cw4, ch4, cbw4, cbh4);
+
+            read_pal_indices(t, pal_idx, b, 1, cw4, ch4, cbw4, cbh4);
+
             if DEBUG_BLOCK_INFO(f, t) {
-                printf(
-                    b"Post-uv-pal-indices: r=%d\n\0" as *const u8 as *const libc::c_char,
-                    ts.msac.rng,
-                );
+                println!("Post-uv-pal-indices: r={}", ts.msac.rng);
             }
         }
-        let mut t_dim: *const TxfmInfo = 0 as *const TxfmInfo;
+
+        let mut t_dim: *const TxfmInfo;
         if frame_hdr.segmentation.lossless[b.seg_id as usize] != 0 {
-            b.uvtx = TX_4X4 as libc::c_int as uint8_t;
-            b.c2rust_unnamed.c2rust_unnamed.tx = b.uvtx;
-            t_dim = &*dav1d_txfm_dimensions
-                .as_ptr()
-                .offset(TX_4X4 as libc::c_int as isize) as *const TxfmInfo;
+            b.uvtx = TX_4X4 as uint8_t;
+            *b.tx_mut() = b.uvtx;
+            t_dim = &dav1d_txfm_dimensions[TX_4X4 as usize];
         } else {
-            b.c2rust_unnamed.c2rust_unnamed.tx = dav1d_max_txfm_size_for_bs[bs as usize][0];
+            *b.tx_mut() = dav1d_max_txfm_size_for_bs[bs as usize][0];
             b.uvtx = dav1d_max_txfm_size_for_bs[bs as usize][f.cur.p.layout as usize];
-            t_dim = &*dav1d_txfm_dimensions
-                .as_ptr()
-                .offset(b.c2rust_unnamed.c2rust_unnamed.tx as isize)
-                as *const TxfmInfo;
-            if frame_hdr.txfm_mode as libc::c_uint
-                == DAV1D_TX_SWITCHABLE as libc::c_int as libc::c_uint
-                && (*t_dim).max as libc::c_int > TX_4X4 as libc::c_int
-            {
+            t_dim = &dav1d_txfm_dimensions[b.tx() as usize];
```

Comment:
This is supposed to be a slice I think.  Or just leave it as a raw ptr for now

Reply:
Is it now implicitly converting back to a raw ptr due to the type annotation on `t_dim`?

Reply:
I would make it a slice, though, since it's coming from an array here.

Reply:
Or maybe not?  I need to look closer.  I thought usually this pattern is for a slice.

Reply:
Looking at the C source, I think it's just supposed to be a ref, not a slice.  So nevermind.